pkg/webhooks/mutator/virt-launcher-mutator.go

```diff
@@ -0,0 +1,177 @@
+	cli          client.Client
+	hcoNamespace string
+	decoder      *admission.Decoder
+	logger       logr.Logger
+}
+
+func NewVirtLauncherMutator(cli client.Client, hcoNamespace string) *VirtLauncherMutator {
+	return &VirtLauncherMutator{
+		cli:          cli,
+		hcoNamespace: hcoNamespace,
+		logger:       log.Log.WithName("virt-launcher mutator"),
+	}
+}
+
+func (m *VirtLauncherMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	m.logInfo("Starting virt-launcher mutator handling")
+
+	if req.Operation != admissionv1.Create {
+		m.logInfo("not a pod creation - ignoring")
+		return admission.Allowed(ignoreOperationMessage)
+	}
+
+	launcherPod := &k8sv1.Pod{}
+	err := m.decoder.Decode(req, launcherPod)
+	if err != nil {
+		m.logErr(err, "cannot decode virt-launcher pod")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	originalPod := launcherPod.DeepCopy()
+
+	hco, err := getHcoObject(ctx, m.cli, m.hcoNamespace)
+	if err != nil {
+		m.logErr(err, "cannot get the HyperConverged object")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if err := m.handleVirtLauncherCreation(launcherPod, hco); err != nil {
+		m.logErr(err, "failed handling launcher pod %s", launcherPod.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	allowResponse := m.getAllowedResponseWithPatches(launcherPod, originalPod)
+	m.logInfo("mutation completed successfully for pod %s", launcherPod.Name)
+	return allowResponse
+}
+
+func (m *VirtLauncherMutator) handleVirtLauncherCreation(launcherPod *k8sv1.Pod, hco *v1beta1.HyperConverged) error {
+	cpuRatioStr, cpuRatioExists := hco.Annotations[cpuLimitToRequestRatioAnnotation]
+	memRatioStr, memRatioExists := hco.Annotations[memoryLimitToRequestRatioAnnotation]
+
+	if !cpuRatioExists && !memRatioExists {
+		return nil
+	}
+
+	if cpuRatioExists {
+		err := m.setResourceRatio(launcherPod, cpuRatioStr, k8sv1.ResourceCPU)
+		if err != nil {
+			return err
+		}
+	}
+	if memRatioExists {
+		err := m.setResourceRatio(launcherPod, memRatioStr, k8sv1.ResourceMemory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *VirtLauncherMutator) setResourceRatio(launcherPod *k8sv1.Pod, ratioStr string, resourceName k8sv1.ResourceName) error {
+	ratio, err := strconv.ParseFloat(ratioStr, 64)
+	if err != nil {
+		return fmt.Errorf("cann't parse %s to float: %v", ratioStr, err)
+	}
+
+	if ratio <= 0 {
+		return fmt.Errorf("negative ratio (%v) is not supported", ratio)
+	}
+
+	for i, container := range launcherPod.Spec.Containers {
+		request, requestExists := container.Resources.Requests[resourceName]
+		limit, limitExists := container.Resources.Limits[resourceName]
+
+		if requestExists && !limitExists {
+			launcherPod.Spec.Containers[i].Resources.Limits[resourceName] = m.multiplyResource(request, ratio)
+		}
+		if !requestExists && limitExists {
+			launcherPod.Spec.Containers[i].Resources.Requests[resourceName] = m.multiplyResource(limit, 1/ratio)
```

Comment:
there is no need to set resource requests if only limits are provided. When users provide only the limits it means that requests will be equal to limits.